modules/linear_boltzmann_solvers/a_lbs_solver/lbs_solver.cc

```diff
@@ -2100,180 +2100,329 @@ LBSSolver::ReadRestartData(const std::string& folder_name, const std::string& fi
 }
 
 void
-LBSSolver::WriteGroupsetAngularFluxes(const LBSGroupset& groupset, const std::string& file_base)
+LBSSolver::WriteAngularFluxes(const std::vector<std::vector<double>>& src,
+                              const std::string& file_base) const
 {
+  // Open the file
   std::string file_name = file_base + std::to_string(Chi::mpi.location_id) + ".data";
-
-  // Open file
   std::ofstream file(file_name,
                      std::ofstream::binary |  // binary file
                        std::ofstream::out |   // no accidental reading
-                       std::ofstream::trunc); // clear file contents when opened
+                       std::ofstream::trunc); // clear contents first
+  ChiLogicalErrorIf(not file.is_open(), "Failed to open " + file_name + ".");
+  Chi::log.Log() << "Writing angular flux to " << file_base;
+
+  // Write the header
+  const int num_bytes = 500;
+  std::string header_info = "OpenSn LinearBoltzmann::Angular flux file\n"
+                            "Header size: " +
+                            std::to_string(num_bytes) +
+                            " bytes\n"
+                            "Structure(type-info):\n"
+                            "uint64_t      num_local_nodes\n"
+                            "unsigned int  num_angles\n"
+                            "unsigned int  num_groups\n"
+                            "Each record:\n"
+                            "  uint64_t      cell_global_id\n"
+                            "  unsigned int  node\n"
+                            "  unsigned int  angle\n"
+                            "  unsigned int  group\n"
+                            "  double        value\n";
 
-  // Check file is open
-  if (not file.is_open())
+  int header_size = (int)header_info.length();
+
+  char header_bytes[num_bytes];
+  memset(header_bytes, '-', num_bytes);
+  strncpy(header_bytes, header_info.c_str(), std::min(header_size, num_bytes - 1));
+  header_bytes[num_bytes - 1] = '\0';
+
+  file << header_bytes;
+
+  // Write macro info
+  auto NODES_ONLY = chi_math::UnknownManager::GetUnitaryUnknownManager();
+
+  const uint64_t num_local_nodes = discretization_->GetNumLocalDOFs(NODES_ONLY);
+  const unsigned int num_groupsets = groupsets_.size();
+
+  file.write((char*)&num_local_nodes, sizeof(uint64_t));
```

Comment:
is the local of nodes in need of such a large size? just wondering about the different sizes of int we use.

Reply:
I recommend to use 64bit numbers consistently through out the whole code for things like this. 
1. We run on 64 bit architectures these days. 
2. It will prevent problems with narrowing conversions.

Reply:
The number of local nodes is the number of local spatial discretization nodes, or number of cells times nodes per cell. It can be quite big. Something like `num_cell_nodes` would be small.